xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp

```diff
@@ -0,0 +1,220 @@
+{
+  if (buffer && dynamic_cast<CVideoBufferDRMPRIME*>(buffer))
+    return new CRendererDRMPRIMEGLES();
+
+  return nullptr;
+}
+
+bool CRendererDRMPRIMEGLES::Register(CWinSystemGbmGLESContext *winSystem)
+{
+  VIDEOPLAYER::CRendererFactory::RegisterRenderer("drm_prime_gles", CRendererDRMPRIMEGLES::Create);
+  m_pWinSystem = winSystem;
+
+  return true;
+}
+
+bool CRendererDRMPRIMEGLES::Configure(const VideoPicture &picture, float fps, unsigned int orientation)
+{
+  for (auto &texture : m_DRMPRIMETextures)
+    texture.Init(m_pWinSystem->GetEGLDisplay());
+
+  return CLinuxRendererGLES::Configure(picture, fps, orientation);
+}
+
+void CRendererDRMPRIMEGLES::ReleaseBuffer(int index)
+{
+  m_DRMPRIMETextures[index].Unmap();
+  CLinuxRendererGLES::ReleaseBuffer(index);
+}
+
+void CRendererDRMPRIMEGLES::AddVideoPicture(const VideoPicture& picture, int index, double currentClock)
+{
+  YUVBUFFER &buf = m_buffers[index];
+  if (buf.videoBuffer)
+    buf.videoBuffer->Release();
+  buf.videoBuffer = picture.videoBuffer;
+  buf.videoBuffer->Acquire();
+}
+
+CRenderInfo CRendererDRMPRIMEGLES::GetRenderInfo()
+{
+  CRenderInfo info;
+  info.max_buffer_size = NUM_BUFFERS;
+  info.optimal_buffer_size = NUM_BUFFERS;
+  info.opaque_pointer = static_cast<void*>(this);
+  return info;
+}
+
+bool CRendererDRMPRIMEGLES::CreateTexture(int index)
+{
+  YUVBUFFER &buf = m_buffers[index];
+  YuvImage &im = buf.image;
+  YUVPLANE &plane = buf.fields[0][0];
+
+  DeleteTexture(index);
+
+  std::memset(&im, 0, sizeof(im));
+  std::memset(&plane, 0, sizeof(YUVPLANE));
+  im.height = m_sourceHeight;
+  im.width  = m_sourceWidth;
+  im.cshift_x = 1;
+  im.cshift_y = 1;
+
+  plane.id = 1;
+
+  return true;
+}
+
+void CRendererDRMPRIMEGLES::DeleteTexture(int index)
+{
+  ReleaseBuffer(index);
+
+  YUVBUFFER &buf = m_buffers[index];
+  buf.fields[0][0].id = 0;
+}
+
+bool CRendererDRMPRIMEGLES::UploadTexture(int index)
+{
+  YUVBUFFER &buf = m_buffers[index];
+
+  CVideoBufferDRMPRIME *buffer = dynamic_cast<CVideoBufferDRMPRIME*>(buf.videoBuffer);
+
+  if (!buffer)
+  {
+    CLog::Log(LOGNOTICE, "CRendererDRMPRIMEGLES::%s - no buffer", __FUNCTION__);
+    return false;
+  }
+
+  m_DRMPRIMETextures[index].Map(buffer);
+
+  YUVPLANE &plane = buf.fields[0][0];
+
+  auto size = m_DRMPRIMETextures[index].GetTextureSize();
+  plane.texwidth  = size.Width();
+  plane.texheight = size.Height();
+  plane.pixpertex_x = 1;
+  plane.pixpertex_y = 1;
+
+  plane.id = m_DRMPRIMETextures[index].GetTexture();
+
+  CalculateTextureSourceRects(index, 1);
+
+  return true;
+}
+
+bool CRendererDRMPRIMEGLES::LoadShadersHook()
+{
+  CLog::Log(LOGNOTICE, "Using DRMPRIMEGLES render method");
+  m_textureTarget = GL_TEXTURE_2D;
+  m_renderMethod = RENDER_DRMPRIME;
+  return true;
+}
+
+bool CRendererDRMPRIMEGLES::RenderHook(int index)
+{
+  CRenderSystemGLES *renderSystem = dynamic_cast<CRenderSystemGLES*>(CServiceBroker::GetRenderSystem());
+  if (!renderSystem)
+    return false;
```

Comment:
If this condition is violated, it's really bad and someone screwed up the code, right? If so, I would vouch for hard-crashing here - either `assert` or even better throw an exception. This kind of stuff is very annoying to debug if it ever goes wrong and you don't know where (does the return value of the function even get used?). We should hard-crash way more often :-)